### DISCUSSION TITLE
Are we allowed to include extra built-in headers?

### DISCUSSION THREAD

**Discussion by hy2850** [ANLAB-KAIST/KENSv3#25] [Project Q&A 1. Application Layer (Socket)] [answered]
The skeleton code in the document clearly dictates : `do not use global variables and do not define/use functions`. Then, how about including extra built-in headers?

For example, network related header like `netdb.h`, or C++ headers for extra functions and classes.
Would like to know if the rule applies to all 4 PAs or only to PA1.

```C
#include <netdb.h>

struct addrinfo *listp;
getaddrinfo(...);
getnameinfo(...);
```

Thank you!

**Comment by Pusnow** [comment, marked as answer]
### getaddrinfo, getnameinfo

They are DNS-related system calls.
They might be okay, but we do not recommend to use them (we will use only IP addresses for grading).

##  C++ headers for extra functions and classes.

They may also be fine. However, some functions, especially socket-related system calls (getsockopt, etc.), do not work.

All functions are acceptable as long as test cases pass.

**Reply by stdeguzman** [reply]
So does this mean that the socket functions listed in the PA #1 document are enough to implement this PA successfully? Or are there other functions we need to use or implement?

**Reply by Pusnow** [reply]
@stdeguzman Yes, the functions (socket, bind, listen, accept, read, write, close, getsockname, getpeername) are enough. For your convenience you may want to use [`inet_addr`](https://man7.org/linux/man-pages/man3/inet.3.html) or [`inet_ntop`](https://man7.org/linux/man-pages/man3/inet_ntop.3.html).
